### PR TITLE
Short Weierstrass wrapping curve for BLS12-377

### PR DESCRIPTION
**Pull request by shamatar**
This EIP adds specification to support an "outer" wrapping curve to allow aggregation of proof and signatures on BLS12-377 (Zexe curve).

### CODE REVIEW

File: EIPS/eip-2541.md [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+
+Also referred as "zero point". For curves with `B != 0` point with coordinates `(0, 0)` (formal zeroes in Fp or Fp3) is *not* on the curve, so encoding of such point `(0, 0)` is used as a convention to encode point of infinity.
+
+##### Boolean encoding for subgroup checks:
+
+For subgroup checks it's required to encode whether it's required to run a subgroup check for the G1/G2 point or not. For this we encode a boolean as a *single byte* `0x00` for `false` and `0x01` for `true`.
+
+##### Encoding of scalars for multiplication operation:
+
+Group order has 377 bits so scalars for multiplication operation is redundantly encoded as `48` bytes by performing BigEndian encoding of the corresponding (unsigned) integer. Corresponding integer is **not** required to be less than or equal than main subgroup size.
```
Comment (iAmMichaelConnor) [inline EIPS/eip-2541.md:108]: @shamatar to be compatible with `32`-byte 'words' of the EVM/Solidity (and to allow easy `mload` and `mstore` operations of 32-byte values in assembly), perhaps scalars should be encoded as `64` bytes instead?
Either:
- the first `16`-bytes could be _enforced_ to be all zeros; or 
- any 64-byte integer could be permitted, and precompiles would need to coerce any `64`-byte value into an element of `Fr` (where `r` is the 377-bit main subgroup order).
(I prefer the first option)

It'd be great to get your thoughts.

I'm facing the same decision making process for an EIP for the BW6-761 curve (which also has a 377-bit scalar field).

File: EIPS/eip-2541.md [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+discussions-to: https://github.com/ethereum/EIPs/pull/2541
+status: Draft
+type: Standards Track
+category: Core
+created: 2020-02-27
+requires : 1109, 2046
+---
+
+<!--You can leave these HTML comments in your merged EIP and delete the visible duplicate text guides, they will not appear and may be helpful to refer to if you edit it again. This is the suggested template for new EIPs. Note that an EIP number will be assigned by an editor. When opening a pull request to submit your EIP, please use an abbreviated title in the filename, `eip-draft_title_abbrev.md`. The title should be 44 characters or less.-->
+
```
Comment (MicahZoltu) [inline EIPS/eip-2541.md:14]:
```suggestion
```

File: EIPS/eip-2541.md [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+## Simple Summary
+<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the EIP.-->
+
```
Comment (MicahZoltu) [inline EIPS/eip-2541.md:17]:
```suggestion
```

File: EIPS/eip-2541.md [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+This precompile adds operation on the SW6 bis curve generated by EY as a precompile in a set necessary to *efficiently* perform operations such as BLS signature verification and perform SNARKs verifications. This curve has a subgroup order equal to the base field of BLS12-377 curve (that in it's term has high 2-addicity). This allow to create zero-knowledge proofs that e.g. aggregate BLS signatures over BLS12-377 curve or can be used in privacy solutions as described in Zexe paper.
+
+## Abstract
+<!--A short (~200 word) description of the technical issue being addressed.-->
+
```
Comment (MicahZoltu) [inline EIPS/eip-2541.md:22]:
```suggestion
```

File: EIPS/eip-2541.md [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+- G2ADD - to perform point addition on a curve twist defined over quadratic extension of the base field
+- G2MUL - to perform point multiplication on a curve twist defined over quadratic extension of the base field
+- G2MULTIEXP - to perform multiexponentiation on a curve twist defined over quadratic extension of the base field
+- PAIRING - to perform a pairing operations between a set of *pairs* of (G1, G2) points
+
+Multiexponentiation operation is included to efficiently aggregate public keys or individual signer's signatures during BLS signature verification, as well as public inputs in SNARKs.
+
+## Motivation
+<!--The motivation is critical for EIPs that want to change the Ethereum protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the EIP solves. EIP submissions without sufficient motivation may be rejected outright.-->
+
```
Comment (MicahZoltu) [inline EIPS/eip-2541.md:37]:
```suggestion
```

File: EIPS/eip-2541.md [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+Motivation of this precompile is to add a cryptographic primitive that allows to get 120+ bits of security for operations and allow aggregation of zero-knowledge proofs over BLS12-377.
+
+## Specification
+<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (go-ethereum, parity, cpp-ethereum, ethereumj, ethereumjs, and [others](https://github.com/ethereum/wiki/wiki/Clients)).-->
+
```
Comment (MicahZoltu) [inline EIPS/eip-2541.md:42]:
```suggestion
```

File: EIPS/eip-2541.md [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+- Multiplication by the unnormalized scalar `(scalar + group_order) * P = scalar * P`
+
+Required properties for pairing operation:
+- Degeneracy `e(P, 0*Q) = e(0*P, Q) = 1` 
+- Bilinearity `e(a*P, b*Q) = e(a*b*P, Q) = e(P, a*b*Q)` (internal test, not visible through ABI)
+
+Test vector for all operations are expanded in this [gist](https://gist.github.com/shamatar/c2cf7608fb4dee4d23f40461da15ad8c) until it's final.
+
+## Implementation
+<!--The implementations must be completed before any EIP is given status "Final", but it need not be completed before the EIP is accepted. While there is merit to the approach of reaching consensus on the specification and rationale before writing code, the principle of "rough consensus and running code" is still useful when it comes to resolving many discussions of API details.-->
```
Comment (MicahZoltu) [inline EIPS/eip-2541.md:261]:
```suggestion
```

File: EIPS/eip-2541.md [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+There is a various choice of existing implementations:
+- EY's `zk-swap-libff` 
+- EIP1962 code bases with fixed parameters
+
+## Security Considerations
+<!--All EIPs must contain a section that discusses the security implications/considerations relevant to the proposed change. Include information that might be important for security discussions, surfaces risks and can be used throughout the life cycle of the proposal. E.g. include security-relevant design decisions, concerns, important discussions, implementation-specific guidance and pitfalls, an outline of threats and risks and how they are being addressed. EIP submissions missing the "Security Considerations" section will be rejected. An EIP cannot proceed to status "Final" without a Security Considerations discussion deemed sufficient by the reviewers.-->
```
Comment (MicahZoltu) [inline EIPS/eip-2541.md:267]:
```suggestion
```

File: EIPS/eip-2541.md [hunk 9]
```diff
@@ -0,0 +1,273 @@
+---
+eip: 2541
+title: EY SW6-Bis curve operations
+author: Alex Vlasov (@shamatar)
+discussions-to: https://github.com/ethereum/EIPs/pull/2541
```
Comment (MicahZoltu) [inline EIPS/eip-2541.md:5]: This cannot be the Pull Request.  You can create a thread on https://ethereum-magicians.org/ or you can create an issue here on GitHub.

**Review by MicahZoltu — CHANGES_REQUESTED** [review]
Just need to remove the comments and pick a new discussions-to location.

File: EIPS/eip-2541.md [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+status: Draft
+type: Standards Track
+category: Core
+created: 2020-02-27
+requires : 1109, 2046
```
Comment (axic) [inline EIPS/eip-2541.md:10]:
```suggestion
requires: 1109, 2046
```

**Pull request closed without merging**